계산기 [STEP1] Hisop

Hi-sop opened the pull request:
@juun97 
안녕하세요! 코드스타터때 뵈었던거같은데 시간 참 빠른거같아요.
이번 프로젝트도 잘 부탁드립니다.

## 고민했던 점
- 배열의 경우 dequeue를 실행할 때 인덱스를 한칸씩 앞으로 당겨주어야 한다고 생각했고 이는 비효율적이라 판단해 연결리스트로 만들어 보고자 했습니다.
- enqueue를 실행할 때 매번 연결의 끝이 어딘지 찾아 newNode를 연결시켜주어야 했으므로 tail을 추가해주었습니다.

## 조언을 얻고 싶었던 점
- Test를 먼저 작성하고 코드를 짜긴 했으나.. TTD를 제대로 하고 있는게 맞는지 확신이 들지 않습니다.
- 연결리스트가 dequeue를 함에 있어서 시간복잡도에서 이득이 있는건 이해했습니다. 하지만 계산의 횟수를 생각하면 간단하게 구현할 수 있는 배열로 만들어도 괜찮았을것 같다는 생각이 듭니다.

Commit:
feat: ItemNode 작성

Commit:
feat: Queue 초안 구성, 연결리스트로 구현해보기

Commit:
feat: enqueue 테스트 작성 및 구현

Commit:
feat: dequeue 테스트 작성 및 구현

Commit:
feat: isEmpty 테스트 작성 및 구현

Commit:
feat: clear 테스트 작성 및 구현

Commit:
feat: count 테스트 작성 및 구현

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,71 @@
+   
+}
+
+class CalculateItemNode: CalculateItem {
+   var value: Double
+   var operatorType: OperatorType
+   var next: CalculateItemNode?
```
1. 현재 만들어진 노드는 Double 타입에만 사용할 수 있겠네요. 어떻게 하면 범용적으로 쓰일 수 있을까요?

2. CalculateItem 은 결국 하나의 아이템을 의미할텐데, Node 에서 핵심이 되는 Item 이 무엇인가요? 

3. 일반적으로 Node 는 다음과 같은 속성을 갖습니다.

- 데이터 요소 (Data Element): Node가 나타내는 값 또는 데이터입니다. 이것은 연결 리스트에 저장되는 실제 정보입니다. 정수, 문자열, 사용자 정의 객체 또는 다른 데이터 형식이 될 수 있습니다.

- 다음 노드 참조 (Next Node Reference): 연결 리스트의 핵심 개념 중 하나로, 다음 Node를 가리키는 참조입니다. 이 링크를 통해 연결 리스트 내의 다음 요소로 이동할 수 있습니다. 맨 마지막 노드의 경우 보통 nil로 초기화됩니다.

하지만 현재 히섭이 만든 데이터 요소는 한 가지가 아닌 두 가지네요.  이렇게 만들어진 노드는 재활용이 불가능 할 것 같습니다. 범용적으로 사용할 수 있게 수정을 해볼까요?

4. CalculateItemNode는 상속이 필요한 Class 인가요?

5. class 로 작성해주신 이유가 있을까요?

6. Node 가 외부에 공개되어야 할 필요가 있나요? Linked List 를 위한 타입이라면 어떻게 처리할 수 있을까요?

Hi-sop replied:
1. 제네릭 문법을 사용해서 수정해볼거같습니다. 지금은 숫자 입력 + 연산자를 한 묶음으로 생각하고 작성했습니다.
2. 사용자가 입력한 값 혹은 클릭한 버튼에 따른 연산자라고 생각합니다. 프로퍼티 하나로 관리하게 수정해볼것 같아요.
3. 넵 수정해볼게요
4. 상속은 가정하지 않았습니다. 수정시에 final키워드를 붙여야겠네요
5. next에 CalculateItemNode? 타입을 지정하기 위해 썼습니다.
6. LinkedList에 속하게 변경할 수 있을것 같습니다.

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,71 @@
+   }
+}
+
+struct CalculateItemQueue {
```
Queue 를 Struct 로 구현하신 이유가 있을까요?

Hi-sop replied:
상속을 가정하지 않았고, 참조타입을 사용하지 않아도 괜찮다는 생각으로 작성하였습니다. 바꿀 이유가 생긴다면 얼마든지 수정할 수 있다고 생각합니다.

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,71 @@
+    var head: CalculateItemNode?
+    var tail: CalculateItemNode?
+    
+    mutating func enqueue(value: Double, operatorType: OperatorType) {
```
mutating 은 어떤 키워드 일까요? 

위에 노드에서 설명드렸다 싶이 주입받고 있는 데이터요소가 2개라 범용적으로 쓰이기엔 힘들 것 같네요.

Hi-sop replied:
프로퍼티를 변경할 가능성이 있는 메서드입니다. 데이터는 수정해보겠습니다

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,71 @@
+        return head == nil
+    }
+    
+    func count() -> Int {
+        var count = 0
+        var currentNode = head
+        
+        while currentNode != nil {
+            currentNode = currentNode?.next
+            count += 1
+        }
+        
+        return count
+    }
```
count 를 실행할 떄마다 연결되어 있는 노드를 모두 검색하는 과정이 필요하겠네요.
이 경우 O(N) 의 시간 복잡도를 가지게 됩니다. 어떻게 하면 O(1) 의 시간복잡도를 가지게 할 수 있을까요?

Hi-sop replied:
큐에 count프로퍼티를 선언해놓고 변동될때마다 관리해주는 아이디어밖에 생각나지 않네요..

juun97 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,78 @@
+        sut = nil
+    }
+
+    func test_enqueue() {
+        sut.enqueue(value: 1, operatorType: .plus)
+        sut.enqueue(value: 2, operatorType: .minus)
+        
+        XCTAssertEqual(sut.head?.value, 1)
+        XCTAssertEqual(sut.head?.operatorType, .plus)
+
+        XCTAssertEqual(sut.tail?.value, 2)
+        XCTAssertEqual(sut.tail?.operatorType, .minus)
+    }
```
테스트 케이스의 네이밍은 어떤 값이 주어졌을 때, 어떤 상황이 오면, 어떤 결과가 나온다.  위 항목들이 설명이 되어야 합니다.
현재는 무엇을 테스트 하려는지 한 눈에 파악하기가 힘듭니다.

- give : 어떤 값이 주어 졌을 때
- when: 어떤 상황이 일어 났을 때
- then : 결과가 어떻게 나오는가

주석을 사용해 3가지 상황을 나눠서 작성해보시면 테스트 코드의 파악이 더 수월해 집니다.

Hi-sop replied:
give when then의 구조로 명확하게 네이밍 수정해볼게요. 'isEmpty실행시_리스트가_비어있을경우_True가_반환되는지'
같은 느낌으로 변경해보겠습니다

juun97 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,78 @@
+        XCTAssertEqual(sut.head?.operatorType, .minus)
+    }
+    
+    func test_isEmpty() {
+        var result = sut.isEmpty()
+        var expectation = true
+        XCTAssertEqual(result, expectation)
+        
+        sut.enqueue(value: 1, operatorType: .plus)
+        result = sut.isEmpty()
+        expectation = false
+        XCTAssertEqual(result, expectation)
+    }
```
하나의 테스트 케이스에선 하나의 테스트만을 진행해야 합니다.
또한 XCTAssertEqual 뿐만이 아니라 다양한 테스트 메서드를 제공하고 있으니 한번 알아보시겠어요?

Hi-sop replied:
두 케이스 다 isEmpty에 속하는 케이스라 판단해서 작성했습니다. 분리하거나 둘중 하나만 남기는 쪽으로 수정하겠습니다.

juun97 reviewed (commented):
계산기 STEP1 고생하셨습니다 히섭 😃
LinkedList 를 시도해본점 매우 훌륭합니다 👍🏻
다만 범용성의 부분에서는 고려가 되지 않은 것 같습니다. 

## 고민했던 점
> 배열의 경우 dequeue를 실행할 때 인덱스를 한칸씩 앞으로 당겨주어야 한다고 생각했고 이는 비효율적이라 판단해 연결리스트로 만들어 보고자 했습니다.

네 맞습니다. Array 로 Queue 를 구현하실 경우 removeFirst 로 dequeue를 할경우 인덱스의 이동이 필요하므로 O(N) 의 시간 복잡도를 가지고 있습니다.

> enqueue를 실행할 때 매번 연결의 끝이 어딘지 찾아 newNode를 연결시켜주어야 했으므로 tail을 추가해주었습니다.

네 맞습니다. 노드들이 연결되어 있다 해서 위와 같은 자료구조를 Linked List 라고 합니다.
여기서 연결의 방향이 단방향인지, 양방향인지에 따라 구분이 되기도 합니다.

## 조언을 얻고 싶었던 점
>Test를 먼저 작성하고 코드를 짜긴 했으나.. TDD를 제대로 하고 있는게 맞는지 확신이 들지 않습니다.

가장 확실한 확인방법은 코드 커버리지를 확인하는 겁니다. 테스트를 할 때 성공하는 테스트도 중요하지만 의도된 실패를 테스트 하는것도 매우 중요합니다.

> 연결리스트가 dequeue를 함에 있어서 시간복잡도에서 이득이 있는건 이해했습니다. 하지만 계산의 횟수를 생각하면 간단하게 구현할 수 있는 배열로 만들어도 괜찮았을것 같다는 생각이 듭니다.

그렇죠 개발자는 항상 만약의 상황을 생각해야 합니다. 만약 해당 큐를 사용해 10만개 이상의 숫자를 저장하는 경우 Array로 만들어진 Queue 와 Linked List로 만들어진 Queue 성능의 차이점은 뚜렷하게 날겁니다.

Commit:
refactor: 코멘트에 따른 수정사항 반영

테스트케이스 네이밍 및 내용 수정
Queue에 제네릭 문법을 적용하여 범용적으로 쓰일 수 있게 함
headNode의 value를 확인하는 peek()추가
countItem프로퍼티를 만들어 count호출마다 모든 노드를 순회하지 않도록 함

Commit:
refactor: 제네릭을 item에만 적용하도록 변경, Queue는 Any로 변경

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,67 @@
+}
+
+struct CalculateItemQueue {
+    var head: CalculateItemNode<Any>?
+    var tail: CalculateItemNode<Any>?
+    var itemCount = 0
+    
+    final class CalculateItemNode<T>: CalculateItem {
+       var value: T
+       var next: CalculateItemNode?
+       
+       init (value: T) {
+           self.value = value
+       }
+    }
```
사용자에게 있어서 Node 는 불필요한 존재입니다. 사용자는 그저 큐의 인터페이스를 이용을하고 큐내부에서 노드를 이용해 자체적으로 List 를 관리를 하게 됩니다.  

현재 Node에서 제너릭을 사용을 해주고 있지만 유저가 Node의 제너릭 모양을 결정할 수가 없는 상태입니다. Queue 가지고 있는 Node 들이 Any 타입이기 때문에 우리가 정해둔 타입외의 어떤 타입이든 올 수 있기에 휴먼에러의 위험성이 매우 높습니다.

그렇다면 어떤 타입이든 일관적으로 관리해줄 수 있는 Queue 를 만드려면 어떻게 해야할까요? 

추가적으로 CalculateItem 를 채택한 타입만을 관리하기 위한 Queue 를 만드려면 해당 프로토콜을 어디서 채택해야 할까요?

Hi-sop replied:
Queue를 제네릭으로 사용하도록 변경했고, 추가적으로 where T: CalculateItem을 통해 프로토콜을 채택한 타입만 들어올 수 있도록 했습니다.

Hi-sop replied:
Queue가 직접 head, tail등의 노드를 관리하지 않도록 리스트 클래스를 분리했습니다.

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,67 @@
+       }
+    }
+    
+    mutating func enqueue(value: Any) {
```
위에서 언급했다 싶이 Any 타입이라면  Double, Int, Float, String 등 어떤 타입이 와도 enqueue 는 작동이 됩니다. 컬렉션 element 들의 요소가 다 달라질 수 있다는 소리입니다. 사용자는 어떤 값을 enqueue 해야할지 모르므로 휴먼에러의 위험성이 매우 높습니다.

Hi-sop replied:
Step2를 보기 전에 구현해서 연산자와 Double값이 차례대로 쌓이는 구조의 큐를 생각하고 Any를 채택했었습니다...

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,67 @@
+        return head == nil
+    }
+    
+    func count() -> Int {
+        return itemCount
+    }
```
itemCount 라는 프로퍼티를 이용함으로써 O(1) 의 시간복잡도를 가지게 됐네요 👍🏻

juun97 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,73 @@
+        XCTAssertEqual(result as? Int, expectation)
+    }
+
+    func test_enqueue로_1_2_3을_넣고_dequeue실행시_head가2인지() {
+        let expectation = 2
+        
+        sut.enqueue(value: 1)
+        sut.enqueue(value: 2)
+        sut.enqueue(value: 3)
+        let _ = sut.dequeue()
+        let result = sut.peek()
+        
+        XCTAssertEqual(result as? Int, expectation)
+    }
```
성공에 대한 테스트가 있다면 실패에 대한 테스트도 해주시는게 좋습니다 🙆🏻‍♂️

Hi-sop replied:
dequeue가 실패했을때 nil인지 테스트하는 항목을 추가했습니다.

juun97 reviewed (commented):
범용적으로 사용가능한 큐를 고민하신 점은 매우 칭찬드립니다 👍🏻
하지만 아직 제너릭을 어떻게 적용을 해야하는지 다소 어색함이 있는 것 같네요!
테스트도 성공에 대한 테스트가 있다면 실패에 대한 테스트도 존재해야 합니다.

코멘트 드린 부분 수정완료되면 머지하도록 하겠습니다.

Commit:
refactor: 코멘트에 따른 수정사항 반영 및 큐, 리스트 분리 등

테스트케이스 수정, 주석추가
Any로 관리하던 큐 제네릭으로 수정
큐에서 관리하던 list의 동작들을 list클래스가 수행하도록 변경
item프로토콜을 채택한 타입만 들어오도록 변경, 따라서 Int extension이 필요했음

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,96 @@
+extension Int: CalculateItem { }
+
+struct CalculateItemList<T> {
+    final private class CalculateItemNode: CalculateItem {
```
이제 큐에서 받는 제너릭 타입은 CalculateItem 을 채택하고 있기대문에 Node에선 채택을 하고 있지 않아도 사용이 가능할 것 같습니다

Hi-sop replied:
아 맞네요 한번 걸러주고 있어서 제외시켜도 되는부분이네요!

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,96 @@
+
+extension Int: CalculateItem { }
+
+struct CalculateItemList<T> {
```
현재 사용하고 있는 Linked List 에서 Calculate 만을 위한 추가적인 기능이 존재하지 않음으로 common 하게 적용될 수 있는 네이밍이 더 적절해 보입니다

juun97 replied:
List 또한 Queue 를 구현하기 위한 요소일 뿐 Queue가 아니기에 파일 분리를 해주세요.
연관성을 주고싶다면 그룹을 묶는 것도 하나의 방법입니다.

Hi-sop replied:
네이밍 변경, 파일 분리했습니다.

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+struct CalculateItemQueue<T> where T: CalculateItem {
```
```suggestion
struct CalculateItemQueue<T: CalculateItem> {
```

 이렇게도 사용 가능합니다

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,96 @@
+        list.appendNode(value: value)
+    }
+    
+    @discardableResult
```
이 어노테이션은 어떤 키워드인가요?

Hi-sop replied:
해당 메서드의 리턴값을 사용하지 않아도 warning이 나오지 않게 합니다. dequeue만을 실행할 때 불편하여 추가해주었던 부분입니다

juun97 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,96 @@
+//  Created by Hisop on 2023/10/05.
+//
+
+protocol CalculateItem { }
+
+extension Int: CalculateItem { }
```
CalculatorItemQueue 파일과는 상관없는 코드인것 같네요.
간단하더라도 분리를 습관화 해주세요

Hi-sop replied:
테스트과정에서 쓰이는 코드라 판단하여 테스트쪽으로 옮겨주었습니다.

juun97 reviewed (commented):
제너릭을 좀 더 제너릭답게 사용할 수 있게 되었네요! 마지막으로 수정 진행되면 머지하겠습니다 🙆🏻‍♂️

Commit:
refactor: List파일 분리, 네이밍 변경

테스트를 위한 Int의 extension코드 테스트쪽으로 이동
